examples/task.hpp

```diff
@@ -23,84 +23,273 @@
+  struct default_task_context_impl::awaiter_context<ParentPromise> {
+    using stop_token_t = std::execution::stop_token_of_t<ParentPromise&>;
+    using stop_callback_t =
+      typename stop_token_t::template callback_type<forward_stop_request>;
+
+    explicit awaiter_context(
+        default_task_context_impl& self, ParentPromise& parent)
+        // Register a callback that will request stop on this basic_task's
+        // stop_source when stop is requested on the parent coroutine's stop
+        // token.
+      : stop_callback_{
+          std::execution::get_stop_token(parent),
+          forward_stop_request{stop_source_}} {
+      self.stop_token_ = stop_source_.get_token();
     }
-    void await_resume() const noexcept {
+
+    std::in_place_stop_source stop_source_{};
+    stop_callback_t stop_callback_;
+  };
+
+// If the parent coroutine's type has a stop token of type in_place_stop_token,
+// we don't need to register a stop callback.
+template <stop_token_provider ParentPromise>
+    requires std::same_as<
+        std::in_place_stop_token,
+        std::execution::stop_token_of_t<ParentPromise&>>
+  struct default_task_context_impl::awaiter_context<ParentPromise> {
+    explicit awaiter_context(
+        default_task_context_impl& self, ParentPromise& parent) {
+      self.stop_token_ = std::execution::get_stop_token(parent);
     }
   };
-  // In a base class so it can be specialized when T is void:
+
+// If the parent coroutine's stop token is unstoppable, there's no point
+// forwarding stop tokens or stop requests at all.
+template <stop_token_provider ParentPromise>
+    requires std::unstoppable_token<
+        std::execution::stop_token_of_t<ParentPromise&>>
+  struct default_task_context_impl::awaiter_context<ParentPromise> {
+    explicit awaiter_context(default_task_context_impl&, ParentPromise&) {}
+  }; 
+
+// Finally, if we don't know the parent coroutine's promise type, assume the
+// worst and save a type-erased stop callback.
+template<>
+  struct default_task_context_impl::awaiter_context<void> {
+    explicit awaiter_context(default_task_context_impl& self, auto&) {}
+
+    template <stop_token_provider ParentPromise>
+      explicit awaiter_context(
+          default_task_context_impl& self, ParentPromise& parent) {
+        // Register a callback that will request stop on this basic_task's
+        // stop_source when stop is requested on the parent coroutine's stop
+        // token.
+        using stop_token_t = std::execution::stop_token_of_t<ParentPromise&>;
+        using stop_callback_t =
+          typename stop_token_t::template callback_type<forward_stop_request>;
+        stop_callback_.emplace<stop_callback_t>(
+            std::execution::get_stop_token(parent),
+            forward_stop_request{stop_source_});
+        self.stop_token_ = stop_source_.get_token();
+      }
+
+    std::in_place_stop_source stop_source_{};
+    std::any stop_callback_{};
```

Comment:
I'm ambivalent, but lean toward the current design. It would seem a little odd for me if we have all the machinery in place _and_ the parent promise is a `stop_token_provider`, but we ignore its stop token. You're right, it could potentially save an allocation (`std::any` could be replaced with something that stores small objects _in situ_), but if you care about that you can simply not define `get_stop_token` for your promise type.